File: 2025/go/main.go
package main

import (
	"fmt"
	"os"
	"time"

	day1 "github.com/DimRev/advent-of-code/day-1"
	day2 "github.com/DimRev/advent-of-code/day-2"
	day3 "github.com/DimRev/advent-of-code/day-3"
	day4 "github.com/DimRev/advent-of-code/day-4"
	day5 "github.com/DimRev/advent-of-code/day-5"
	day6 "github.com/DimRev/advent-of-code/day-6"
	day7 "github.com/DimRev/advent-of-code/day-7"
	day8 "github.com/DimRev/advent-of-code/day-8"
	day9 "github.com/DimRev/advent-of-code/day-9"
	"github.com/DimRev/advent-of-code/lib"
)

type Command string

const (
	D1P1 Command = "d1p1"
	D1P2 Command = "d1p2"

	D2P1 Command = "d2p1"
	D2P2 Command = "d2p2"

	D3P1 Command = "d3p1"
	D3P2 Command = "d3p2"

	D4P1 Command = "d4p1"
	D4P2 Command = "d4p2"

	D5P1 Command = "d5p1"
	D5P2 Command = "d5p2"

	D6P1 Command = "d6p1"
	D6P2 Command = "d6p2"

	D7P1 Command = "d7p1"
	D7P2 Command = "d7p2"

	D8P1 Command = "d8p1"
	D8P2 Command = "d8p2"

	D9P1 Command = "d9p1"
	D9P2 Command = "d9p2"
)

var commands = []Command{D1P1, D1P2}

var cmdMap = map[Command]func(){
	D1P1: day1.Day1Part1,
	D1P2: day1.Day1Part2,

	D2P1: day2.Day2Part1,
	D2P2: day2.Day2Part2,

	D3P1: day3.Day3Part1,
	D3P2: day3.Day3Part2,

	D4P1: day4.Day4Part1,
	D4P2: day4.Day4Part2,

	D5P1: day5.Day5Part1,
	D5P2: day5.Day5Part2,

	D6P1: day6.Day6Part1,
	D6P2: day6.Day6Part2,

	D7P1: day7.Day7Part1,
	D7P2: day7.Day7Part2,

	D8P1: day8.Day8Part1,
	D8P2: day8.Day8Part2,

	D9P1: day9.Day9Part1,
	D9P2: day9.Day9Part2,
}

func main() {
	if len(os.Args) < 2 {
		fmt.Println("Usage: go run main.go <command>")
		fmt.Println("Available commands:")
		for _, cmd := range commands {
			fmt.Printf("\t- %s\n", cmd)
		}
		os.Exit(1)
	}

	cmdStr := os.Args[1]
	cmd := Command(cmdStr)

	if _, ok := cmdMap[cmd]; !ok {
		fmt.Printf("Unknown command: %s\n", cmdStr)
		fmt.Println("Available commands:")
		for _, cmdEnum := range commands {
			fmt.Printf("\t- %s\n", cmdEnum)
		}
		os.Exit(1)
	}

	startTs := time.Now()
	cmdMap[cmd]()
	endTs := time.Now()
	elapsed := endTs.Sub(startTs)
	fmt.Printf("Finished running %s in %d(μs)\n", cmdStr, elapsed.Microseconds())

	err := lib.PopulateRenderer(cmdStr, int(elapsed.Microseconds()))
	if err != nil {
		fmt.Printf("Error populating renderer: %v\n", err)
	}
}

File: 2025/python/main.py
import sys
import time
from enum import Enum
from typing import Callable, Dict

from day_1 import day1_part1, day1_part2
from day_2 import day2_part1, day2_part2
from day_3 import day3_part1, day3_part2
from day_4 import day4_part1, day4_part2
from day_5 import day5_part1, day5_part2
from day_6 import day6_part1, day6_part2
from utils import populate_renderer

class Command(str, Enum):
    D1P1 = "d1p1"
    D1P2 = "d1p2"
    D2P1 = "d2p1"
    D2P2 = "d2p2"
    D3P1 = "d3p1"
    D3P2 = "d3p2"
    D4P1 = "d4p1"
    D4P2 = "d4p2"
    D5P1 = "d5p1"
    D5P2 = "d5p2"
    D6P1 = "d6p1"
    D6P2 = "d6p2"


CMD_MAP: Dict[Command, Callable[[], None]] = {
    Command.D1P1: day1_part1,
    Command.D1P2: day1_part2,
    Command.D2P1: day2_part1,
    Command.D2P2: day2_part2,
    Command.D3P1: day3_part1,
    Command.D3P2: day3_part2,
    Command.D4P1: day4_part1,
    Command.D4P2: day4_part2,
    Command.D5P1: day5_part1,
    Command.D5P2: day5_part2,
    Command.D6P1: day6_part1,
    Command.D6P2: day6_part2,
}


def main():
    if len(sys.argv) < 2:
        print("Usage: python main.py <command>")
        print("Available commands:")
        for cmd in Command:
            print(f"\t- {cmd.value}")
        sys.exit(1)

    cmd_str = sys.argv[1]

    try:
        cmd = Command(cmd_str)
    except ValueError:
        print(f"Unknown command: {cmd_str}")
        print("Available commands:")
        for cmd_enum in Command:
            print(f"\t- {cmd_enum.value}")
        sys.exit(1)

    start_ts = time.perf_counter_ns()
    CMD_MAP[cmd]()
    end_ts = time.perf_counter_ns()
    elapsed = (end_ts - start_ts) / 1000
    print(f"Finished running {cmd_str} in {elapsed} (μs)")
    populate_renderer(cmd_str, elapsed)

if __name__ == "__main__":
    main()